### PR TITLE
Remove intermediate heap allocations when interacting with Base58 addresses

### PR DESCRIPTION
**Pull request by c410-f3r**
## No heap allocations

Taking into consideration that an address has an small length upper bond, it is then possible to use stack allocations to remove the overhead of intermediate heap allocations when creating or parsing Base58 addresses. Local tests show an improvement of ~40% compared to the current code.

## Standalone crate

I am trying to create a custom lightweight library that communicates with nodes (something like subxt) and `Ss58AddressFormat` is the source of truth that tells if a parachain/network has indeed a Substrate Base58 address prefix. So knowing each network prefix will be handy for future development of this library or if I am not mistaking, almost a must for proper inter-connections.

That said, it is really hard and sad for me to import `sp-core` just to see ~250 additional transients dependencies in my `Cargo.lock` file and suffer more than three additional minutes of wasted compilation time, thus the separation of `Ss58AddressFormat` and related structures into a standalone crate.

### CODE REVIEW

**Comment by xlc** [comment]
> almost a must for proper inter-connections.

Why do you say that?

ss58 prefix is not really used in runtime currently and not involved by XCM at all. It is currently only used by frontend and by looking at the prefix, there is still not possible to know parachain networks we are dealing with because networks could be sharing prefix. e.g. Statemine is using same prefix as Kusama.

**Comment by c410-f3r** [comment]
> > almost a must for proper inter-connections.
> 
> Why do you say that?
> 
> ss58 prefix is not really used in runtime currently and not used by XCM at all. By only looking at the prefix, it is still not possible to know which network we are dealing with because networks could be sharing prefix. e.g. Statemine is using same prefix as Kusama.

This was I perception I had while developing the library. Maybe now such thing isn't necessary but could be useful in future

File: client/cli/src/commands/utils.rs [hunk 1]
```diff
@@ -202,11 +210,11 @@ where
 	match output {
 		OutputType::Json => {
 			let json = json!({
-				"networkId": String::from(network_override),
+				"networkId": network_override.to_string(),
 				"publicKey": format_public_key::<Pair>(public_key.clone()),
-				"ss58PublicKey": public_key.to_ss58check_with_version(network_override),
+				"ss58PublicKey": public_key.to_ss58check_with_version(network_override).as_str(),
```
Comment (gilescope) [inline client/cli/src/commands/utils.rs:215]: What's the advantage of `.as_str()`? Is the Display implementation currently inefficient?

Comment (c410-f3r) [reply client/cli/src/commands/utils.rs:215]: IIRC, the `json!` macro expects types that can be deferred into some trait/type/struct used by `serde_json`. `ArrayString` isn't supported but `&str` is.

**Comment by bkchr** [comment]
This will be solved by: https://github.com/paritytech/substrate/issues/9696

**Comment by c410-f3r** [comment]
I guess the 40% performance improvement is not worth it

**Comment by bkchr** [comment]
@c410-f3r only the performance improvements would be okay, but you changed tons of other stuff in this pr as well. If you open a new one, with just the performance improvements, fine.

**Comment by gilescope** [comment]
I'm just puzzled slightly why `.to_string()` reduces allocations... Mostly when I think of `.to_string() -> String` I feel it often forces a new allocation.

**Comment by c410-f3r** [comment]
> @c410-f3r only the performance improvements would be okay, but you changed tons of other stuff in this pr as well. If you open a new one, with just the performance improvements, fine.

Done -> #9756

**Pull request closed without merging**